aws/smithy-aws-traits/src/main/java/software/amazon/smithy/aws/traits/endpointdiscovery/CleanDiscoveryTraitTransformer.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.smithy.aws.traits.endpointdiscovery;
+
+import java.util.Collection;
+import java.util.Set;
+import java.util.stream.Collectors;
+import software.amazon.smithy.model.Model;
+import software.amazon.smithy.model.shapes.ServiceShape;
+import software.amazon.smithy.model.shapes.Shape;
+import software.amazon.smithy.model.shapes.ShapeId;
+import software.amazon.smithy.model.traits.ErrorTrait;
+import software.amazon.smithy.model.traits.Trait;
+import software.amazon.smithy.model.transform.ModelTransformer;
+import software.amazon.smithy.model.transform.ModelTransformerPlugin;
+
+/**
+ * Removes the endpoint discovery trait from a service if the referenced operation or error are removed.
+ */
+public class CleanDiscoveryTraitTransformer implements ModelTransformerPlugin {
```

Comment:
The other `ModelTransformerPlugin` implementations handle all relevant changes to make sure the resulting model is in a valid state. Could this do so via `filterTraits` on the other traits? Should it?

Reply:
I think they are related but separate concerns. By decorating an operation with `discoveredEndpoint` you're saying that it needs to have an endpoint discovered for it to work (or at least work optimally). I don't think that changes if you project away the operation used for discovery.